DRAFT - 22.0.0.3-beta blog post (#2031)
OpenLiberty/blogs: posts/2022-02-17-beta-release-post.adoc

---
layout: post

> title: "TITLE"

Reviewer: ```suggestion
title: "Concurrency 3.0 for Jakarta EE 10 in Open Liberty 22.0.0.3-beta"
```

author_github: https://github.com/ryan-storey
seo-title: Concurrency 3.0 for Jakarta EE 10 in Open Liberty 22.0.0.3-beta - OpenLiberty.io
seo-description: Open Liberty 22.0.0.3-beta offers Concurrency 3.0 for Jakarta EE 10, which introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
blog_description: Open Liberty 22.0.0.3-beta offers Concurrency 3.0 for Jakarta EE 10, which introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Concurrency 3.0 for Jakarta EE 10 in Open Liberty 22.0.0.3-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 22.0.0.3-beta offers Concurrency 3.0 for Jakarta EE 10, which introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.

Reviewer: ```suggestion
It's only been a couple of short months since we announced Jakarta EE 9.1 support in Open Liberty 21.0.0.12, and we're already looking forward as we collaborate on and begin to implement Jakarta EE 10 features.  Open Liberty 22.0.0.3-beta offers our first Jakarta EE 10 beta feature, Concurrency 3.0.  Concurrency 3.0 introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
```

        try {
            // Resource reference lookup requires context of the application component,
            DataSource db = InitialContext.doLookup("java:comp/env/jdbc/account-db-ref");
            ...
            return Asynchronous.Result.complete(overdueAccounts);
        } catch (NamingException | SQLException x) {
            throw new CompletionException(x);
        }
    }

    ...
----

When the application invokes the asynchronous method, the container intercepts the invocation and hands back a completion stage representing its eventual completion, which is arranged by the `ManagedExecutorService`.

[source, java]
----
    public void doGet(HttpServletRequest req, HttpServletResponse res) {
        ...
        accounts.findOverdue(60).thenAccept(this::sendReminder);
        ...
    }

    void sendReminder(List<Account> overdue) {
        try {
            // Resource reference lookup requires context of the application component,
            DataSource db = InitialContext.doLookup("java:comp/env/jdbc/customer-db-ref");
            ...
        } catch (NamingException | SQLException x) {
            throw new CompletionException(x);
        }
    }
----

> The Concurrency 3.0 specification hasn't been published yet because Jakarta EE 10 is still under development. However, you can preview a draft of it in asciidoc link:https://github.com/eclipse-ee4j/concurrency-api/blob/2abe74ce30589ef53e248009fcb446491fd42a27/specification/src/main/asciidoc/jakarta-concurrency.adoc[here].

Reviewer: ```suggestion
The Concurrency 3.0 specification hasn't been published yet because Jakarta EE 10 is still under development. However, you can preview a draft of it at the link:https://github.com/eclipse-ee4j/concurrency-api/blob/2abe74ce30589ef53e248009fcb446491fd42a27/specification/src/main/asciidoc/jakarta-concurrency.adoc[concurrency-api GitHub page].
```

> Similarly, the API jar will not be available on Maven until the specification is released. In the mean time, a draft copy of it is made available on DHE at the link:http://public.dhe.ibm.com/ibmdl/export/pub/software/olrepo/io/openliberty/jakarta/enterprise/concurrent/jakarta.enterprise.concurrent-api/3.0.0.20220120/[following location].

Reviewer: ```suggestion
Similarly, the API jar will not be available on Maven until the specification is released, but in the mean time, you can check out the link:http://public.dhe.ibm.com/ibmdl/export/pub/software/olrepo/io/openliberty/jakarta/enterprise/concurrent/jakarta.enterprise.concurrent-api/3.0.0.20220120/[draft copy.]
```

            <url>https://public.dhe.ibm.com/ibmdl/export/pub/software/olrepo</url>
        </repository>
    </repositories>
    <dependencies>
        <dependency>
            <groupId>io.openliberty.jakarta.enterprise.concurrent</groupId>
            <artifactId>jakarta.enterprise.concurrent-api</artifactId>
            <version>3.0.0.20220120</version>
        </dependency>
    <dependencies>
----

=== Try it now 

To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 15, Java SE 11, or Java SE 8.

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
  <groupId>io.openliberty.beta</groupId>
  <artifactId>openliberty-runtime</artifactId>
  <version>22.0.0.3-beta</version>
  <type>pom</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty.beta', name: 'openliberty-runtime', version: '[22.0.0.3-beta,)'
}
----

Or take a look at our link:{url-prefix}/downloads/#runtime_betas[Downloads page].

> [#jakarta]
> == Jakarta EE 9 Features
> 
> As of 21.0.0.12 the Jakarta EE 9.1 features are now out of beta and fully supported. Check out the link:https://openliberty.io/blog/2021/11/26/jakarta-ee-9.1.html[release blog post] for these features.
> 
> === Try it now
> 
> To try out these Jakarta EE 9 features on Open Liberty in a lightweight package, just update your build tools to pull the Open Liberty Jakarta EE 9 Beta Features package instead of the main release. The beta works with Java SE 15, Java SE 11, or Java SE 8.
> 
> If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:
> 
> [source,xml]
> ----
> <dependency>
>     <groupId>io.openliberty.beta</groupId>
>     <artifactId>openliberty-jakartaee9</artifactId>
>     <version>22.0.0.3-beta</version>
>     <type>zip</type>
> </dependency>
> ----
> 
> Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:
> 
> [source,gradle]
> ----
> dependencies {
>     libertyRuntime group: 'io.openliberty.beta', name: 'openliberty-jakartaee9', version: '[22.0.0.3-beta,)'
> }
> ----
> 
> Or take a look at our link:{url-prefix}/downloads/#runtime_betas[Downloads page].
> 

Reviewer: ```suggestion
```
